### PR TITLE
Implement client interceptors in C# (DRAFT)

### PR DESCRIPTION
**Pull request by mehrdada**

### CODE REVIEW

**Comment by mehrdada** [comment]
@jtattermusch please take a look to ensure the general direction for client-side interceptors is acceptable.

**Comment by jtattermusch** [comment]
Is there a corresponding gRFC document  that summarizes the idea or you wanted to do some hands-on experiments first?

**Comment by mehrdada** [comment]
@jtattermusch per your advice, I thought I'd show you this API for sanity check before posting an official gRFC. I specifically debated a bit whether to make a new interface for `IInterceptor` or reuse the `CallInvoker` base class for this purpose, which would have required some way to ensure the underlying `CallInvoker` in the chain to be given to the interceptor object somehow and I thought it's not cleanly possible to do so.

File: src/csharp/Grpc.Core/InterceptorBase.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+using System;
+
+namespace Grpc.Core
+{
+    public abstract class InterceptorBase : IInterceptor
+    {
+        protected virtual object BeforeCall<TRequest, TResponse>(ref Method<TRequest, TResponse> method, ref string host, ref CallOptions options, TRequest request) { return null; }
+
+        protected virtual void AfterCall<TResponse>(object state, TResponse response) { }
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/InterceptorBase.cs]: what's the state for?

**Comment by mehrdada** [comment]
abandoned in favor of https://github.com/grpc/grpc/pull/12613

**Pull request closed without merging**